llarp/dns.cpp

```diff
@@ -124,10 +124,180 @@ code_domain(char *&buffer, const std::string &domain) throw()
+
+  bytes.push_back(domain.size() - start);  // last label length octet
+  for(size_t i = start; i < domain.size(); i++)
+  {
+    bytes.push_back(domain[i]);  // last label octets
+    // llarp::LogInfo("Writing ", domain[i], " at ", i);
+  }
+  bytes.push_back(0);  // end it
+}
+
+// expects host order
+void
+vput16bits(std::vector< byte_t > &bytes, uint16_t value) throw()
+{
+  char buf[2]        = {0};
+  char *write_buffer = buf;
+  htobe16buf(write_buffer, value);
+  bytes.push_back(buf[0]);
+  bytes.push_back(buf[1]);
+}
+
+// expects host order
+void
+vput32bits(std::vector< byte_t > &bytes, uint32_t value) throw()
+{
+  char buf[4]        = {0};
+  char *write_buffer = buf;
+  htobe32buf(write_buffer, value);
+  bytes.push_back(buf[0]);
+  bytes.push_back(buf[1]);
+  bytes.push_back(buf[2]);
+  bytes.push_back(buf[3]);
+}
+
+void
+dns_writeType(std::vector< byte_t > &bytes, llarp::dns::record *record)
+{
+  llarp::dns::type_1a *type1a = dynamic_cast< llarp::dns::type_1a * >(record);
+  if(type1a)
+  {
+    std::vector< byte_t > more_bytes = type1a->to_bytes();
+    llarp::LogDebug("[1]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+
+  llarp::dns::type_2ns *type2ns =
+      dynamic_cast< llarp::dns::type_2ns * >(record);
+  if(type2ns)
+  {
+    std::vector< byte_t > more_bytes = type2ns->to_bytes();
+    llarp::LogDebug("[2]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+
+  llarp::dns::type_5cname *type5cname =
+      dynamic_cast< llarp::dns::type_5cname * >(record);
+  if(type5cname)
+  {
+    std::vector< byte_t > more_bytes = type5cname->to_bytes();
+    llarp::LogDebug("[5]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+
+  llarp::dns::type_12ptr *type12ptr =
+      dynamic_cast< llarp::dns::type_12ptr * >(record);
+  if(type12ptr)
+  {
+    std::vector< byte_t > more_bytes = type12ptr->to_bytes();
+    llarp::LogDebug("[12]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+  llarp::dns::type_15mx *type15mx =
+      dynamic_cast< llarp::dns::type_15mx * >(record);
+  if(type15mx)
+  {
+    std::vector< byte_t > more_bytes = type15mx->to_bytes();
+    llarp::LogDebug("[15]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+  llarp::dns::type_16txt *type16txt =
+      dynamic_cast< llarp::dns::type_16txt * >(record);
+  if(type16txt)
+  {
+    std::vector< byte_t > more_bytes = type16txt->to_bytes();
+    llarp::LogDebug("[15]Adding ", more_bytes.size());
+    bytes.insert(bytes.end(), more_bytes.begin(), more_bytes.end());
+  }
+}
+
+std::vector< byte_t >
+packet2bytes(dns_packet &in)
+{
+  std::vector< byte_t > write_buffer;
+  vput16bits(write_buffer, in.header->id);
+
+  int fields = (in.header->qr << 15);   // QR => message type, 1 = response
```

Comment:
hoist these magic numbers to constants somewhere?

Reply:
Geez uhm, I'm not even sure I understand what these are. Basically just some position adjustment, not really anything labelled.